[Cutlass] Implement EVT example tensor creation

mlazos opened the pull request:
This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150910
* #150909
* #150908
* #150907
* #151406
* #150906
* #151713
* #151405
* #150905
* __->__ #150904

udpates to example tensor creation

Previously merged:
* https://github.com/pytorch/pytorch/pull/150903
* https://github.com/pytorch/pytorch/pull/150346
* https://github.com/pytorch/pytorch/pull/150345
* https://github.com/pytorch/pytorch/pull/150344

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

mlazos committed:
[Cutlass] Implement EVT example tensor creation

udpates to example tensor creation

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos commented:
I'm going to add a quick unit test for this

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/codegen/cuda/cutlass_lib_extensions/evt_extensions.py:
```diff
@@ -41,13 +42,80 @@
+    }
+
+    def _check_contiguous(shape: tuple[int, ...], stride: tuple[int, ...]) -> bool:
+        if len(shape) != len(stride):
+            return False
+
+        running_product = 1
+        for shape_i, stride_i in zip(shape, stride):
+            if shape_i == 1:
+                continue
+            if stride_i != running_product:
+                return False
+            running_product *= shape_i
+
+        return True
```
`ir.is_contiguous_strides_for_shape` ?

eellison commented on torch/_inductor/codegen/cuda/cutlass_lib_extensions/evt_extensions.py:
```diff
@@ -41,13 +42,80 @@
+        write_names: list[str],
+        buffer_renames: dict[str, str],
+        name_to_buffer: dict[str, Union[ComputedBuffer, InputBuffer]],
+    ):
```
+1 to return type

eellison commented on torch/_inductor/codegen/cuda/cutlass_lib_extensions/evt_extensions.py:
```diff
@@ -41,13 +42,80 @@
+        example_tensors = {}
+
+        def cutlass_tensor_from_buffer(buffer: Union[ComputedBuffer, InputBuffer]):
+            shape = tuple(int(x) for x in buffer.get_layout().size)
+            stride = tuple(int(x) for x in buffer.get_layout().stride)
+
+            is_column_major = _check_contiguous(shape, stride)
+            is_row_major = _check_contiguous(shape[::-1], stride[::-1])
```
int(symint) will throw e.g. if the symint is unbacked. We should be able to check this without converting the symints -> int. 

`buffer.get_layout().is_transposed` and `buffer.get_layout().is_contiguous`

mlazos replied:
I do need to know whether row or column major though unfortunately (another suggestion for NVIDIA)

mlazos replied:
for the symint thing, I also need them to be ints though for cutlass

eellison commented on torch/_inductor/codegen/cuda/cutlass_lib_extensions/evt_extensions.py:
```diff
@@ -41,13 +42,80 @@
+            is_column_major = _check_contiguous(shape, stride)
+            is_row_major = _check_contiguous(shape[::-1], stride[::-1])
+
+            if not is_row_major and not is_column_major:
+                raise RuntimeError(
+                    f"Cannot create example tensor for {buffer.get_name()} with \
+non-contiguous layout, recieved stride: {stride} and shape: {shape}"
+                )
```
Do we check these conditions before attempting ? or are we expecting to recover the error?

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

Previously merged:
* https://github.com/pytorch/pytorch/pull/150903
* https://github.com/pytorch/pytorch/pull/150346
* https://github.com/pytorch/pytorch/pull/150345
* https://github.com/pytorch/pytorch/pull/150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

Previously merged:
* https://github.com/pytorch/pytorch/pull/150903
* https://github.com/pytorch/pytorch/pull/150346
* https://github.com/pytorch/pytorch/pull/150345
* https://github.com/pytorch/pytorch/pull/150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

Previously merged:
* https://github.com/pytorch/pytorch/pull/150903
* https://github.com/pytorch/pytorch/pull/150346
* https://github.com/pytorch/pytorch/pull/150345
* https://github.com/pytorch/pytorch/pull/150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge

henrylhtsang commented on torch/_inductor/codegen/cuda/cutlass_lib_extensions/evt_extensions.py:
```diff
@@ -41,13 +46,74 @@
 
     _CUTLASS_C_DTYPES = OrderedSet(dtype2ctype.values())  # type: ignore[var-annotated]
 
+    TORCH_TO_CUTLASS_DTYPE = {
```
we have torch_dtype_to_cutlass_type in cutlass_utils.py

henrylhtsang replied:
wait you don't even use this

mlazos replied:
Yeah I can remove, I think I left it even after I switched the util

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "Sorry but this is breaking the test_example_tensor_creation test internally. See D73519195 for more details. To validate your fixes internally, you can follow the instructions here: https://fburl.com/fixing-ghfirst-reverts"

mlazos committed:
Update on "[Cutlass] Implement EVT example tensor creation"

This PR implements a translation layer from inductor IR to "example tensors" the expected arguments of the EVT tracer. These tensors basically store the name, shape, stride, and dtype of the tensor and allow an ast-based python parse to generate the EVT C++.


udpates to example tensor creation

Previously merged:
* https://github.com/pytorch/pytorch/pull/150903
* https://github.com/pytorch/pytorch/pull/150346
* https://github.com/pytorch/pytorch/pull/150345
* https://github.com/pytorch/pytorch/pull/150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge
